### ISSUE TITLE
Support the Rule keyword

### ISSUE THREAD

**Issue by ciaranmcnulty** [Behat/Gherkin#140]
This is new in Gherkin v6

https://docs.cucumber.io/gherkin/reference/#rule

**Comment by stof** [comment]
If we want to support it, we should also consider refactoring Behat and Gherkin to separate the Gherkin AST from what they call pickles. This way, the Behat runner would not have to know about the extra grouping features of the Gherkin syntax.
This would be a BC break for hooks (some hooks would have to disappear, as the execution itself would not know about outlines or backgrounds anymore for instance).

In any case, adding that would require a new behat/gherkin major version, as adding rules inside features would break Behat when running them.

**Comment by ciaranmcnulty** [comment]
Yes, if we went that route it might make sense to semi-generate the parser the same way that's done in other languages (and this repo might become a compat later)

Or cucumber-engine will become the future and we won't need a Gherkin parser for behat...

**Comment by stof** [comment]
Well, what other languages are doing is not semi-generating the parser. They are calling the go parser (as a CLI) and then only rebuilding the output objects. I'm not sure that's what we want for now.

**Comment by benztest** [comment]
I was wondering if there was any further updates regarding the Rule keyword?

As in my recent attempt to include two rules in a single feature results in an error. 
``` 
  [Behat\Gherkin\Exception\ParserException]
  Expected Step, but got text: "Rule: Authenticate as a learner" in file: C:\<path>\<file>.feature
```
.feature:
```
Feature: ...

Rule: Rule: Do no authenticate

Scenario: ...

Scenario: ...

Rule: Rule: Authenticate as a learner

Scenario: ...

Scenario: ...
```

**Comment by ciaranmcnulty** [comment]
Uncommenting these lines will now produce a failing test for the Rule keyword implementation

https://github.com/Behat/Gherkin/blob/master/tests/Behat/Gherkin/Cucumber/CompatibilityTest.php#L29-L30

**Comment by Harrisonbro** [comment]
Sorry if this should be obvious, but is there somewhere we can see how support for the Rule keyword moves from this issue into a version of Behat we can use as end users? Sorry, I’m not familiar enough with the project or how open source works to figure this out myself.

**Comment by ciaranmcnulty** [comment]
This isn’t yet implemented here. Once it is, a new version of Behat will need to add support

**Comment by kferrone** [comment]
Well that was a hard stop. So what's the status here. Years of being open doesn't look too promising. 

This looks hopeful: https://github.com/Behat/Behat/pull/1404
But it's been open for about 4yrs.

**Comment by acoulton** [comment]
Exploring using the official parser is still on the radar for the future, but it is not in the immediate plan. The current focus is on getting our parser to parity with the official parsers (which is now true apart from for Rule) and on getting Behat 4.0 out.

This is also tracked in https://github.com/Behat/Behat/issues/1451 - I have a half-idea of how we might be able to achieve it in a backwards-compatible way (behind our GherkinCompatibilityMode flag).

I would like to try and find time to look at this before we promote the GHERKIN_32 parser mode to being "stable" for the Behat 4.0 release - can't make any promises as it will depend on whether my idea survives contact with reality.

**Comment by acoulton** [comment]
Backwards-compatible Rule support (when parsing in GHERKIN_32 compatibility mode) was implemented in #430 and will be in the next release.
